feat: implement function locators (#10632)
puppeteer/puppeteer: docs/guides/locators.md

# Locators

> Locators is an innovative, experimental API that combines the functionalities of

Reviewer: ```suggestion
Locators is a new, experimental API that combines the functionalities of
```

sounds too pompous, IMO

> waiting and actions into a unified unit. With additional precondition checks, it

Reviewer: ```suggestion
waiting and actions. With additional precondition checks, it
```

"unified unit" does not sound too good, unit is already unified by definition.

:::

## Use cases

### Waiting for an element

```ts
await page.locator('button').wait();
```

The following preconditions are automatically checked:

- Waits for the element to become
  [visible](https://pptr.dev/api/puppeteer.elementhandle.isvisible/) or hidden.

### Waiting for a function

```ts
await page
  .locator(() => {

>     return new Promise(resolve => setTimeout(resolve, 500));

Reviewer: let's write an example that actually locates something? I cannot think of a good example, perhaps, using an Intersection observer or some other API to locate an element based on custom logic?
